QEfficient/cloud/finetune.py

```diff
@@ -0,0 +1,204 @@
+    dataset_train = get_preprocessed_dataset(
+        dataset_processer, dataset_config, split="train", context_length=train_config.context_length
+    )
+
+    dataset_val = get_preprocessed_dataset(
+        dataset_processer, dataset_config, split="test", context_length=train_config.context_length
+    )
+
+    # TODO: vbaddi, check if its necessary to do this?
+    # dataset_train = ConcatDataset(
+    #             dataset_train, chunk_size=train_config.context_length
+    #         )
+    ##
+    train_dl_kwargs = get_dataloader_kwargs(train_config, dataset_train, dataset_processer, "train")
+    print("length of dataset_train", len(dataset_train))
+    custom_data_collator = get_custom_data_collator(dataset_processer, dataset_config)
+    if custom_data_collator:
+        print("custom_data_collator is used")
+        train_dl_kwargs["collate_fn"] = custom_data_collator
+    # Create DataLoaders for the training and validation dataset
+    train_dataloader = torch.utils.data.DataLoader(
+        dataset_train,
+        num_workers=train_config.num_workers_dataloader,
+        pin_memory=True,
+        **train_dl_kwargs,
+    )
+    print(f"--> Num of Training Set Batches loaded = {len(train_dataloader)}")
+
+    eval_dataloader = None
+    if train_config.run_validation:
+        # if train_config.batching_strategy == "packing":
+        #     dataset_val = ConcatDataset(
+        #         dataset_val, chunk_size=train_config.context_length
+        #     )
+
+        val_dl_kwargs = get_dataloader_kwargs(train_config, dataset_val, dataset_processer, "val")
+        if custom_data_collator:
+            val_dl_kwargs["collate_fn"] = custom_data_collator
+
+        eval_dataloader = torch.utils.data.DataLoader(
+            dataset_val,
+            num_workers=train_config.num_workers_dataloader,
+            pin_memory=True,
+            **val_dl_kwargs,
+        )
+        if len(eval_dataloader) == 0:
+            raise ValueError(
+                f"The eval set size is too small for dataloader to load even one batch. Please increase the size of eval set. ({len(eval_dataloader)=})"
+            )
+        else:
+            print(f"--> Num of Validation Set Batches loaded = {len(eval_dataloader)}")
+
+    longest_seq_length, longest_seq_ix = get_longest_seq_length(
+        torch.utils.data.ConcatDataset([train_dataloader.dataset, eval_dataloader.dataset])
+    )
+    print(
+        f"The longest sequence length in the train data is {longest_seq_length}, "
+        f"passed context length is {train_config.context_length} and overall model's context length is "
+        f"{model.config.max_position_embeddings}"
+    )
+    model.to(train_config.device)
+    optimizer = optim.AdamW(
+        model.parameters(),
+        lr=train_config.lr,
+        weight_decay=train_config.weight_decay,
+    )
+    scheduler = StepLR(optimizer, step_size=1, gamma=train_config.gamma)
+
+    # wrap model with DDP
+    if train_config.enable_ddp:
```

Comment:
we should shuffle dataset if ddp is enabled

Reply:
addressed this.